designs/full-cluster-lifecycle-api.md

```diff
@@ -0,0 +1,224 @@
+
+We no longer allow workload cluster to have its own GitOpsConfig resource. Users can still specify `cluster.spec.gitOpsRef` which points to the management cluster GitOpsConfig resource if exists.
+
+**A GitOps reconciler might be introduced later to automate the Git operations where the cluster spec file is created/updated in the remote repo.
+
+### CLI
+
+If the user decides to continue using the CLI for lifecycle management, this will just read an EKSA cluster spec from disk and push it to a running cluster where the EKSA controller is running, either a bootstrap (kind) cluster or a management cluster. If the admission webhook returns an error the user will see an error message. If the spec is accepted for processing then the CLI will have to wait for a condition in the status that signals that the cluster has been created (for example, spec.Ready= true), a timeout event and/or a failure reason in the status. It will also download the Kubeconfig necessary to access the newly created workload cluster. This can be done in the same way as it is done right now.
+
+Using the controller to manage self-managed clusters (from the bootstrap cluster) allows us to eliminate most of the complexity introduced by trying to share the same logic for both self-managed and managed clusters. For self-managed clusters the CLI will still need to perform some extra steps, like creating a `kind` cluster, installing CAPI controllers, moving CAPI resources, etc.
+
+### CRD Defaults
+
+Currently, as part of the CLI logic we modify the cluster spec objects provided by the user, adding defaults (eg. vSphere templates) and modifying some fields to have valid values (eg. vSphere paths with wildcards).
+
+Defaults can be added both from the types mutation webhooks or from the reconcilers themselves (depending on the logic needed to come up with those values).
+
+However, we should stop modifying the fields that the user has already specified. When those values need to be transformed (in order to be validated or used), we will run the same logic the CLI currently uses and store the final values in the status object.
+
+### Controller and Self-managed Clusters
+
+Currently, it is only possible to update a few fields through the EKSA controller (like worker node replicas). This includes self-managed clusters.
+
+As part of this design, that support for self-managed clusters will be removed. This simplifies the controller and avoids confusion about the difference in upgrade capabilities between self-managed clusters and managed clusters. The upgrade capabilities through the CLI will remain intact.
+
+We will revisit this once we decide to offer full cluster lifecycle support for self-managed clusters.
+
+## User Experience
+
+For now, the only way to create and upgrade a management cluster is through the CLI. Unlike the previous versions, upgrading a management cluster through GitOps is no longer supported.
```

Comment:
Might be nice to have a sentence about why you cannot upgrade a management cluster through GitOps.

Reply:
Done!